core/simd/util/util.odin

```diff
@@ -0,0 +1,160 @@
+/*
+	(c) Copyright 2024 Feoramund <rune@swevencraft.org>.
+	Made available under Odin's BSD-3 license.
+
+	List of contributors:
+		Feoramund: `index_byte` procedures.
+*/
+
+// package simd_util implements compositions of SIMD operations for optimizing
+// the core library where available.
+package simd_util
+
+import "base:intrinsics"
+
+@private SCAN_WIDTH :: 32
+
+@(private, rodata)
+simd_scanner_indices := #simd[SCAN_WIDTH]u8 {
+	 0,  1,  2,  3,  4,  5,  6,  7,
+	 8,  9, 10, 11, 12, 13, 14, 15,
+	16, 17, 18, 19, 20, 21, 22, 23,
+	24, 25, 26, 27, 28, 29, 30, 31,
+}
+
+/*
+Scan a slice of bytes for a specific byte.
+
+This procedure safely handles slices of any length, including empty slices.
+
+Inputs:
+- data: A slice of bytes.
+- c: The byte to search for.
+
+Returns:
+- index: The index of the byte `c`, or -1 if it was not found.
+*/
+index_byte :: proc "contextless" (data: []u8, c: byte) -> (index: int) #no_bounds_check {
+	length := len(data)
+	i := 0
+
+	// Guard against small strings.
+	if length < SCAN_WIDTH {
+		for /**/; i < length; i += 1 {
+			if data[i] == c {
+				return i
+			}
+		}
+		return -1
+	}
+
+	ptr := cast(int)cast(uintptr)raw_data(data)
+
+	alignment_start := (SCAN_WIDTH - ptr % SCAN_WIDTH) % SCAN_WIDTH
+
+	// Iterate as a scalar until the data is aligned on a `SCAN_WIDTH` boundary.
+	//
+	// This way, every load in the vector loop will be aligned, which should be
+	// the fastest possible scenario.
+	for /**/; i < alignment_start; i += 1 {
+		if data[i] == c {
+			return i
+		}
+	}
+
+	// Iterate as a vector over every aligned chunk, evaluating each byte simultaneously at the CPU level.
+	scanner: #simd[SCAN_WIDTH]u8 = c
+	tail := length - (length - alignment_start) % SCAN_WIDTH
+
+	for /**/; i < tail; i += SCAN_WIDTH {
```

Comment:
It is probably a net-gain on anything but i386 to go wider.  Sticking the wider loops in a `when ODIN_ARCH != .i386` block is still cross-platform.

Reply:
Do you mean changing `SCAN_WIDTH` to something like `64` for non-386 platforms?

Reply:
It's not the only 32-bit platform. I'd think `SCAN_WIDTH :: 8 * size_of(uintptr)` to be a good thing to try.

Reply:
That shouldn't be a problem. There only needs to be an index table for 32 and 64 then.

Reply:
That's another 12% faster on 1 GiB for me.

Reply:
Although that 12% faster was without optimization. The win goes away at `-o:speed`.

`SCAN_WIDTH :: 8 * size_of(uintptr) (-o:speed)`
```
[INFO ] --- [2024-08-10 11:11:36] [benchmark_simd_util.odin:115:benchmark_simd_index_hot()]
        +++      15B | 33ns
 (last) +++      15B | 33ns
        +++      16B | 0s
 (last) +++      16B | 33ns
        +++      17B | 100ns
 (last) +++      17B | 100ns
        +++      31B | 100ns
 (last) +++      31B | 3.4µs
        +++      32B | 100ns
 (last) +++      32B | 66ns
        +++      33B | 66ns
 (last) +++      33B | 100ns
        +++     256B | 33ns
 (last) +++     256B | 66ns
        +++     512B | 100ns
 (last) +++     512B | 0s
        +++  1.00KiB | 0s
 (last) +++  1.00KiB | 100ns
        +++  1.00MiB | 27.8µs
 (last) +++  1.00MiB | 27.166µs
        +++  1.00GiB | 79.691566ms
 (last) +++  1.00GiB | 67.379533ms
 ```
 `SCAN_WIDTH :: 4 * size_of(uintptr) (-o:speed)`
 ```
 [INFO ] --- [2024-08-10 11:14:37] [benchmark_simd_util.odin:115:benchmark_simd_index_hot()]
        +++      15B | 33ns
 (last) +++      15B | 100ns
        +++      16B | 33ns
 (last) +++      16B | 33ns
        +++      17B | 66ns
 (last) +++      17B | 66ns
        +++      31B | 0s
 (last) +++      31B | 66ns
        +++      32B | 0s
 (last) +++      32B | 33ns
        +++      33B | 33ns
 (last) +++      33B | 100ns
        +++     256B | 66ns
 (last) +++     256B | 33ns
        +++     512B | 100ns
 (last) +++     512B | 100ns
        +++  1.00KiB | 100ns
 (last) +++  1.00KiB | 33ns
        +++  1.00MiB | 22.7µs
 (last) +++  1.00MiB | 22.7µs
        +++  1.00GiB | 81.7955ms
 (last) +++  1.00GiB | 55.585ms
 ```

Reply:
On aggressive, the SCANWIDTH of 64 is more clearly a win.
```
[INFO ] --- [2024-08-10 11:17:49] [benchmark_simd_util.odin:115:benchmark_simd_index_hot()]
        +++      15B | 33ns
 (last) +++      15B | 33ns
        +++      16B | 33ns
 (last) +++      16B | 33ns
        +++      17B | 33ns
 (last) +++      17B | 33ns
        +++      31B | 33ns
 (last) +++      31B | 33ns
        +++      32B | 33ns
 (last) +++      32B | 33ns
        +++      33B | 33ns
 (last) +++      33B | 33ns
        +++     256B | 66ns
 (last) +++     256B | 66ns
        +++     512B | 66ns
 (last) +++     512B | 66ns
        +++  1.00KiB | 33ns
 (last) +++  1.00KiB | 33ns
        +++  1.00MiB | 19.433µs
 (last) +++  1.00MiB | 14.266µs
        +++  1.00GiB | 95.351966ms
 (last) +++  1.00GiB | 51.4385ms
 ```
 Versus 32:
 ```
 [INFO ] --- [2024-08-10 11:16:53] [benchmark_simd_util.odin:115:benchmark_simd_index_hot()]
        +++      15B | 100ns
 (last) +++      15B | 66ns
        +++      16B | 33ns
 (last) +++      16B | 33ns
        +++      17B | 66ns
 (last) +++      17B | 33ns
        +++      31B | 33ns
 (last) +++      31B | 33ns
        +++      32B | 0s
 (last) +++      32B | 66ns
        +++      33B | 0s
 (last) +++      33B | 0s
        +++     256B | 33ns
 (last) +++     256B | 66ns
        +++     512B | 100ns
 (last) +++     512B | 66ns
        +++  1.00KiB | 66ns
 (last) +++  1.00KiB | 0s
        +++  1.00MiB | 28.266µs
 (last) +++  1.00MiB | 21.833µs
        +++  1.00GiB | 96.1965ms
 (last) +++  1.00GiB | 51.4338ms
 ```

Reply:
`0s` is an interesting result. All of the results which are evenly divisible by `33` are also interesting. Is this an artifact of timekeeping on Windows?

Reply:
Could be. We can ask for a higher resolution timer for `accurate_sleep` for example. The `0s` also surprised me.

Reply:
0s is because you are using `time.now` instead of `time.tick_now`, `tick_now` is better for timing things (not affected by system time) and is more precise (especially on windows).